google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java

```diff
@@ -2922,6 +3066,42 @@ boolean isValid() {
     }
   }
 
+  /**
+   * Returns a multiplexed session. We would always return the session which is at the front of the
+   * queue since this will be the most recently created session.
+   */
+  SessionFuture getMultiplexedSessionWithFallback() throws SpannerException {
+    if (options.getUseMultiplexedSession()) {
+      ISpan span = tracer.getCurrentSpan();
+      span.addAnnotation("Acquiring multiplexed session");
+      MultiplexedSessionWaiterFuture waiter = null;
+      synchronized (lock) {
+        if (resourceNotFoundException != null) {
+          span.addAnnotation("Database has been deleted");
+          throw SpannerExceptionFactory.newSpannerException(
+              ErrorCode.NOT_FOUND,
+              String.format(
+                  "The session has been invalidated because a previous RPC returned 'Database not found': %s",
+                  resourceNotFoundException.getMessage()),
+              resourceNotFoundException);
+        }
+        MultiplexedSession session = multiplexedSession.get();
+        if (session != null) {
+          span.addAnnotation("Acquired multiplexed session", "sessionId", session.getName());
+          incrementNumSessionsInUse(true);
+        } else {
+          span.addAnnotation("No multiplexed session available. Adding to waiter queue.");
+          maybeCreateMultiplexedSession();
+          waiter = new MultiplexedSessionWaiterFuture();
+          multiplexedSessionWaiters.add(waiter);
+        }
+        return checkoutMultiplexedSession(span, session, waiter);
```

Comment:
I think that all of this could be replaced by a simple `return currentMultiplexedSessionFuture`. That is:
1. The `SessionPool` would have a field of type `AtomicReference<SettableApiFuture<MultiplexedSessions>>` that holds the current multiplexed session. During startup/initial creation, that future is still pending. In all other cases, it will be a future that is done, and that contains the current multiplexed session.
2. This method can then always just return the single instance of that future, instead of having to create additional objects every time a multiplexed session is 'checked out'. AFAICT, the only state that we maintain in `MultiplexedSessionFuture` is the `span`, but that is only used to set the span on the underlying `SessionImpl`, which is always the same instance, meaning that we don't need a separate `MultiplexedSessionFuture` instance for each checkout. (And it also means that we should get rid of the `span` field on that class, because it now does something that is clearly wrong and confusing.)

Reply:
Done, simplified this flow by using `AtomicReference<SettableApiFuture<MultiplexedSessionFuture>>`.